File: contest 5/Closest friends!.py
def closestFriends(arr, n):
    result = []
    stack = []

    for i in range(n):
        while stack and arr[stack[-1]] >= arr[i]:
            stack.pop()
        
        if not stack:
            result.append(-1)
        else:
            result.append(stack[-1])
        
        stack.append(i)
    
    return result

File: contest 5/Closest friends!.cpp
vector<int> closestFriends(int arr[], int n) {
    vector<int> result(n, -1);
    stack<int> s;

    for (int i = 0; i < n; ++i) {
        while (!s.empty() && arr[s.top()] >= arr[i]) {
            s.pop();
        }

        if (!s.empty()) {
            result[i] = s.top();
        }

        s.push(i);
    }

    return result;
}
